### PR TITLE
Add test to validate Keycloak in prod mode, with keystore generation

### PR DESCRIPTION
**Pull request by mamccorm**
> **DRAFT: Debugging some CI issues with the test**

Added new test to validate Keycloak running in production mode. This entails creating a keystore and volume mounting it into a Keycloak container. 

Validates by checking the service is running and that an API call can be made with credentials.

### CODE REVIEW

File: images/keycloak/tests/01-docker-production-mode.sh [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    fi
+}
+
+TEST_keycloak_api_accessible() {
+  local retries=3
+  local delay=5 # seconds
+  local attempt=0
+
+  while [[ $attempt -le $retries ]]; do
+    local response_code=$(curl -k -I -s -o /dev/null -w "%{http_code}" "$KEYCLOAK_URL/realms/master")
```
Comment (stormqueen1990) [inline images/keycloak/tests/01-docker-production-mode.sh:140]: Does it make sense to combine `curl` flags like so?

```suggestion
    local response_code=$(curl -kIso /dev/null -w "%{http_code}" "$KEYCLOAK_URL/realms/master")
```

File: images/keycloak/tests/01-docker-production-mode.sh [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+declare -a missing_terms=()
+
+create_keystore() {  
+  rm -rf "$KEYSTORE_PATH"
+  keytool -v -keystore $KEYSTORE_PATH \
+  -alias $KEYCLOAK_HOSTNAME \
+  -genkeypair -sigalg SHA512withRSA -keyalg RSA \
+  -dname CN=$KEYCLOAK_HOSTNAME \
+  -storepass $KEYSTORE_PASSWORD
```
Comment (stormqueen1990) [inline images/keycloak/tests/01-docker-production-mode.sh:33]: This is really a nit, but I'd indent this a bit for better visualization
```suggestion
    -alias $KEYCLOAK_HOSTNAME \
    -genkeypair -sigalg SHA512withRSA -keyalg RSA \
    -dname CN=$KEYCLOAK_HOSTNAME \
    -storepass $KEYSTORE_PASSWORD
```

File: images/keycloak/tests/01-docker-production-mode.sh [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  docker run \
+  -v "$KEYSTORE_PATH:/usr/share/java/keycloak/conf/server.keystore" \
+  --detach --rm \
+  --name local-keycloak -p $KEYCLOAK_PORT:8443 \
+  -e KEYCLOAK_ADMIN=admin \
+  -e KEYCLOAK_ADMIN_PASSWORD=$KEYSTORE_PASSWORD \
+  "${IMAGE_NAME}" \
+  start \
+  --hostname=$KEYCLOAK_HOSTNAME \
+  --https-key-store-password=$KEYSTORE_PASSWORD
```
Comment (stormqueen1990) [inline images/keycloak/tests/01-docker-production-mode.sh:46]: Same nit as the previous one.
```suggestion
  docker run \
    -v "$KEYSTORE_PATH:/usr/share/java/keycloak/conf/server.keystore" \
    --detach --rm \
    --name local-keycloak -p $KEYCLOAK_PORT:8443 \
    -e KEYCLOAK_ADMIN=admin \
    -e KEYCLOAK_ADMIN_PASSWORD=$KEYSTORE_PASSWORD \
    "${IMAGE_NAME}" \
    start \
    --hostname=$KEYCLOAK_HOSTNAME \
    --https-key-store-password=$KEYSTORE_PASSWORD
```

File: images/keycloak/tests/01-docker-production-mode.sh [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        exit 1
+    fi
+}
+
+TEST_keycloak_api_accessible() {
+  local retries=3
+  local delay=5 # seconds
+  local attempt=0
+
+  while [[ $attempt -le $retries ]]; do
```
Comment (stormqueen1990) [inline images/keycloak/tests/01-docker-production-mode.sh:139]: Could this use arithmetic bash expressions like you did before?
```suggestion
  while (( attempt <= retries )); do
```

File: images/keycloak/tests/01-docker-production-mode.sh [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    fi
+    attempt=$((attempt+1))
+  done
+}
+
+
+TEST_keycloak_api_get_users() {
+    # Validate users can be retrieved from the Keycloak API.
+    local users_json=$(keycloak_api_get_users)
+    local exttracted_username=$(echo "$users_json" | jq -r '.[] | select(.username=="admin") | .username')
```
Comment (stormqueen1990) [inline images/keycloak/tests/01-docker-production-mode.sh:162]: Minor typo

```suggestion
    local extracted_username=$(echo "$users_json" | jq -r '.[] | select(.username=="admin") | .username')
```

File: images/keycloak/tests/01-docker-production-mode.sh (hunk 2 above)
Comment (mamccorm) [reply images/keycloak/tests/01-docker-production-mode.sh:33]: 100%, thanks and also for taking the time to do the inline suggestions

File: images/keycloak/tests/01-docker-production-mode.sh (hunk 5 above)
Comment (mamccorm) [reply images/keycloak/tests/01-docker-production-mode.sh:162]: doh!

**Pull request closed without merging**